src/main/java/ch/njol/skript/effects/EffBuildRestrictions.java

```diff
@@ -0,0 +1,97 @@
+package ch.njol.skript.effects;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.aliases.ItemType;
+import ch.njol.skript.doc.*;
+import ch.njol.skript.lang.Effect;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import com.destroystokyo.paper.Namespaced;
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@Name("Apply Build Restrictions")
+@Description("Allow or prevent an item to destroy or be placed on certain types of blocks while in /gamemode adventure.")
+@Examples("allow player's tool to destroy (stone and oak wood planks)")
+@Since("INSERT VERSION")
+@RequiredPlugins("Paper")
+public class EffBuildRestrictions extends Effect {
+
+	static {
+		if (Skript.methodExists(ItemMeta.class, "setDestroyableKeys")) {
+			Skript.registerEffect(EffBuildRestrictions.class,
+				"allow %~itemtypes% to (:(destroy|break|mine|be placed on)) %itemtypes%",
+				"(disallow|prevent) %~itemtypes% from (:(destroying|breaking|mining|being placed on)) %itemtypes%");
+			// should this require notnull? and can the patterns here be made any better?
+		}
+	}
+
+	private boolean allow;
+	private boolean destroy;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<ItemType> items;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<ItemType> modifyWith;
+	private static final List<String> destroyTags = List.of("destroy", "destroying", "break", "breaking", "mine", "mining");
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		this.items = (Expression<ItemType>) exprs[0];
+		this.modifyWith = (Expression<ItemType>) exprs[1];
+		this.allow = matchedPattern == 0;
+		this.destroy = destroyTags.stream().anyMatch(parseResult::hasTag);
```

Comment:
You only have one place syntax, better to check if it has a tag for placing instead of checking 6 different ones for the breaking.